File: core/src/main/scala/org/onemodel/core/model/AttributeWithValidAndObservedDates.scala
/*  This file is part of OneModel, a program to manage knowledge.
    Copyright in each year of 2014, 2016-2016 inclusive, Luke A Call; all rights reserved.
    OneModel is free software, distributed under a license that includes honesty, the Golden Rule, guidelines around binary
    distribution, and the GNU Affero General Public License as published by the Free Software Foundation, either version 3
    of the License, or (at your option) any later version.  See the file LICENSE for details.
    OneModel is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
    You should have received a copy of the GNU Affero General Public License along with OneModel.  If not, see <http://www.gnu.org/licenses/>

  ---------------------------------------------------
  (See comment in this place in PostgreSQLDatabase.scala about possible alternatives to this use of the db via this layer and jdbc.)
*/
package org.onemodel.core.model

import org.onemodel.core.model.Database

object AttributeWithValidAndObservedDates {
  def getDatesDescription(mValidOnDate:Option[Long], mObservationDate:Long): String = {
    val validDateDescr: String =
      if (mValidOnDate.isEmpty) "unsp'd"
      else if (mValidOnDate.get == 0) "all time"
      else Attribute.usefulDateFormat(mValidOnDate.get)
    val observedDateDescr: String = Attribute.usefulDateFormat(mObservationDate)
    "valid " + validDateDescr + ", obsv'd " + observedDateDescr
  }
}

abstract class AttributeWithValidAndObservedDates(mDB: Database, mId: Long) extends Attribute(mDB, mId) {
  protected def assignCommonVars(parentIdIn: Long, attrTypeIdIn: Long, validOnDateIn: Option[Long], observationDateIn: Long, sortingIndexIn: Long) {
    mValidOnDate = validOnDateIn
    // observationDate is not expected to be None, like mValidOnDate can be. See var def for more info.
    mObservationDate = observationDateIn
    super.assignCommonVars(parentIdIn, attrTypeIdIn, sortingIndexIn)
  }

  def getDatesDescription: String = {
    AttributeWithValidAndObservedDates.getDatesDescription(getValidOnDate, getObservationDate)
  }

  private[onemodel] def getValidOnDate: Option[Long] = {
    if (!mAlreadyReadData) readDataFromDB()
    mValidOnDate
  }

  private[onemodel] def getObservationDate: Long = {
    if (!mAlreadyReadData) readDataFromDB()
    mObservationDate
  }

  /**
   * For descriptions of the meanings of these variables, see the comments
   * on createTables(...), and examples in the database testing code in PostgreSQLDatabase or Database classes.
   */
  protected var mValidOnDate: Option[Long] = None
  protected var mObservationDate: Long = 0L
}

File: core/src/model/attribute_with_valid_and_observed_dates.rs
/*  This file is part of OneModel, a program to manage knowledge.
    Copyright in each year of 2014, 2016-2017 inclusive and 2023, Luke A. Call.
    OneModel is free software, distributed under a license that includes honesty, the Golden Rule,
    and the GNU Affero General Public License as published by the Free Software Foundation;
    see the file LICENSE for license version and details.
    OneModel is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
    You should have received a copy of the GNU Affero General Public License along with OneModel.  If not, see <http://www.gnu.org/licenses/>
*/
use crate::model::attribute::Attribute;
// use crate::util::Util;
use sqlx::{Postgres, Transaction};
use std::cell::{RefCell};
use std::rc::Rc;

// (For more info see "supertraits" in The Book (or in anki notes).)
pub trait AttributeWithValidAndObservedDates: Attribute {
    //%%
    //was:
    // fn assign_common_vars(
    //     parent_id_in: i64,
    //     attr_type_id_in: i64,
    //     valid_on_date_in: Option<i64>,
    //     observation_date_in: i64,
    //     sorting_index_in: i64,
    // );

    // was:
    //  fn assign_common_vars(parent_id_in: i64, attr_type_id_in: i64, valid_on_date_in: Option<i64>, observation_date_in: i64, sorting_index_in: i64) {
    //    valid_on_date = valid_on_date_in
    //    // observation_date is not expected to be None, like valid_on_date can be. See let mut def for more info.;
    //    observation_date = observation_date_in
    //    super.assign_common_vars(parent_id_in, attr_type_id_in, sorting_index_in)
    //  }

    // just call Util directly from callers, instead?:  Except when called for FA??--dift.
    // fn get_dates_description(valid_on_date: Option<i64>, observation_date: i64) -> String {
    //     Util::get_dates_description(valid_on_date, observation_date)
    //   }

    fn get_valid_on_date(&mut self,
        transaction: Option<Rc<RefCell<Transaction<Postgres>>>>,
    ) -> Result<Option<i64>, anyhow::Error>;

    fn get_observation_date(
        &mut self,
        transaction: Option<Rc<RefCell<Transaction<Postgres>>>>,
    ) -> Result<i64, anyhow::Error>;

    // For descriptions of the meanings of these variables, see the comments
    // on create_tables(...), and examples in the database testing code in PostgreSQLDatabase or Database classes.
    // %%put these in the structs implementing this trait, along w/ those above methods!
    // protected let mut valid_on_date: Option<i64> = None;
    // protected let mut observation_date: i64 = 0L;
}
